File: src/cmcandy/Java_Language_Answers/_0035.java
package cmcandy.Java_Language_Answers;

/**
 * Created by CHEN on 2020/5/7.
 */
public class _0035 {

    static class Solution {
        public int searchInsert(int[] nums, int target) {
            if (nums[0]>target) return 0;
            if (nums[nums.length-1]<target) return nums.length;
            return search(nums, 0, nums.length - 1, target);
        }

        public int search(int[] nums, int first, int last, int target) {
            int mid = (first + last) / 2;
            if (first==last-1){
                if (nums[first]==target)return first;
                else if (nums[last]==target) return  last;
                else return last;
            }
            if (nums[mid] > target) return search(nums, first, mid, target);
            else if (nums[mid] < target) return search(nums, mid, last, target);
            else if (nums[mid] == target) return mid;
            return -1;
        }
    }
}

File: src/cmcandy/Python_language_Answers/_0035.py
from typing import List


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:

        # 二分法
        left, right = 0, len(nums) - 1
        if nums[left] >= target:
            return left
        elif nums[right] < target:
            return right + 1
        elif nums[right] == target:
            return right
        while left < right:
            mid = (left + right) // 2
            if nums[mid] > target:
                if nums[mid - 1] < target:
                    return mid
                elif nums[mid - 1] == target:
                    return mid - 1
                else:
                    right = mid - 1
            elif nums[mid] == target:
                return mid
            elif nums[mid] < target:
                if nums[mid + 1] >= target:
                    return mid + 1
                else:
                    left = mid + 1
        return -1



